### ISSUE TITLE
Material for a very soft (no bounce) collision

### ISSUE THREAD

**Issue by spacejack** [schteppe/cannon.js#297]
I want to create an almost bounce-free collision... I've been playing with the ContactMaterial properties but the rebound is still too fast. I want it to be like a ball hitting a pillow or a bean bag.

This is the best I can do so far:

```
const mat = new CANNON.ContactMaterial(
    ballMat, pillowMat,
    {
        friction: 10.0,
        restitution: 0.0,
        contactEquationRelaxation: 10.0,
        frictionEquationStiffness: 1
    }
)
```

Any tips? :)

**Comment by schteppe** [comment]
Experiment with those parameters. You can use this demo to get the effect you want (press space to rerun with same params): http://schteppe.github.io/cannon.js/demos/singleBodyOnPlane.html

Also consider reducing gravity and increasing body damping. This will reduce the overall velocity, therefore the overall contact penetration and therefore also the amount of rebound (this is a common gamedev physics trick, also used to improve physics performance).

**Comment by spacejack** [comment]
Hi schteppe, thanks. You mean the `linearDamping` property when creating a body? That seems to increase "air drag" but that's not what I want.

I did some more experimentation, it seems setting `restitution = 0` and high value for `contactEquationRelaxation` like 1000 or more give me the best result. The other parameters don't seem to change much so I'm omitting them.

If needed I guess I can simply reduce the velocity myself upon contact.

**Comment by spacejack** [comment]
So, experimenting some more, I think I'll get better results if I just clamp the velocity myself. I've run into a bit of a snag however.

I used `addEventListener('collide', ...)` to the pillow object for collisions. It fires when the ball hits it. At that point, I clamp the ball velocity to a small value, however it doesn't seem to change the velocity. I'm guessing because the collision is still happening and my velocity change is overwritten by cannon.

I see in [this thread](https://github.com/schteppe/cannon.js/issues/249) that there should be an `'endContact'` event. I tried using that instead of `'collide'`, however it doesn't seem to fire.

Any ideas on how best to manually adjust the velocity when contact is detected?

Also, I can't seem to find a list of event names that can be used with `Body#addEventListener` - that would be helpful. :)

**Comment by schteppe** [comment]
Right. Yea, damping will reduce velocity of objects in general. Probably not the effect you want, just a suggestion.

You probably want to clamp the velocity after each step. Something like this?

``` js
world.addEventListener("postStep", function(event){
    if(shouldClamp){
        body.velocity.y = 0;
        shouldClamp = false;
    }
});
world.addEventListener("beginContact", function(event){
    shouldClamp = true;
});
```

I'm not sure why event doesn't work for you. Try using the latest master branch version from github?

Looks like the docs are not available for those events yet! Wanna add?

**Comment by spacejack** [comment]
> Looks like the docs are not available for those events yet! Wanna add?

Sure, I could help with that.

Hmm, I used `body.addEventListener('endContact'...` I'll try re-building, I thought I was using a fresh build but I'll double check.

Been pretty busy tring to meet a deadline. I'll try to put together a small test case, there's another little quirk I want to demonstrate as well.

**Comment by spacejack** [comment]
I made a little test - this demonstrates that 'collide' events fire, but 'endContact' events don't:

http://www.spacejack.ca/cannon-contact-example/

I put up a source repo here:
https://github.com/spacejack/cannon-contact-example

This is using a freshly pulled cannon.js repo, built with `grunt`.

I set up the endContact event [on this line in the source](https://github.com/spacejack/cannon-contact-example/blob/master/js/main.js#L113).
